The failure is:
```
 ERROR: testGetFeatureInfoThousandSeparator (__main__.TestQgsServerWMSLocaleOverride)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/QGIS/tests/src/python/test_qgsserver.py", line 160, in tearDown
    self.temporary_dir.cleanup()
AttributeError: 'TestQgsServerWMSLocaleOverride' object has no attribute 'temporary_dir'
```

--- a/tests/src/python/test_qgsserver.py
+++ b/tests/src/python/test_qgsserver.py
@@ -49,6 +49,7 @@ from utilities import unitTestDataPath
 import osgeo.gdal  # NOQA
 import tempfile
 import base64
+from shutil import copytree
 
 
 start_app()
@@ -112,14 +113,47 @@ class QgsServerTestBase(unittest.TestCase):
                 self.assertEqual(expected_values, response_values, msg=msg + "\nXML attribute values differ at line {0}: {1} != {2}".format(line_no, expected_values, response_values))
             line_no += 1
 
-    def setUp(self):
+    @classmethod
+    def setUpClass(self):
         """Create the server instance"""
         self.fontFamily = QgsFontUtils.standardTestFontFamily()
         QgsFontUtils.loadStandardTestFonts(['All'])
 
-        self.testdata_path = unitTestDataPath('qgis_server') + '/'
+        self.temporary_dir = tempfile.TemporaryDirectory()
+        self.temporary_path = self.temporary_dir.name
 
-        d = unitTestDataPath('qgis_server_accesscontrol') + '/'
+        # Copy all testdata to the temporary directory
+        copytree(unitTestDataPath('qgis_server'), os.path.join(self.temporary_path, 'qgis_server'))
+        copytree(unitTestDataPath('qgis_server_accesscontrol'), os.path.join(self.temporary_path, 'qgis_server_accesscontrol'))
+
+        for f in [
+            'empty_spatial_layer.dbf',
+            'empty_spatial_layer.prj',
+            'empty_spatial_layer.qpj',
+            'empty_spatial_layer.shp',
+            'empty_spatial_layer.shx',
+            'france_parts.dbf',
+            'france_parts.prj',
+            'france_parts.qpj',
+            'france_parts.shp',
+            'france_parts.shp.xml',
+            'france_parts.shx',
+            'landsat.tif',
+            'points.dbf',
+            'points.prj',
+            'points.shp',
+            'points.shx',
+            'requires_warped_vrt.tif',
+        ]:
+            os.symlink(
+                unitTestDataPath(f),
+                os.path.join(self.temporary_path, f)
+            )
+
+        self.testdata_path = os.path.join(self.temporary_path, 'qgis_server') + '/'
+
+        d = os.path.join(self.temporary_path, 'qgis_server_accesscontrol')
+
         self.projectPath = os.path.join(d, "project.qgs")
         self.projectAnnotationPath = os.path.join(d, "project_with_annotations.qgs")
         self.projectStatePath = os.path.join(d, "project_state.qgs")
@@ -139,14 +173,17 @@ class QgsServerTestBase(unittest.TestCase):
         # Disable landing page API to test standard legacy XML responses in case of errors
         os.environ["QGIS_SERVER_DISABLED_APIS"] = "Landing Page"
 
-    def tearDown(self):
+    @classmethod
+    def tearDownClass(self):
         """Cleanup env"""
 
-        super().tearDown()
+        super().tearDownClass()
         try:
             del os.environ["QGIS_SERVER_DISABLED_APIS"]
         except KeyError:
             pass
+
+        self.temporary_dir.cleanup()
 
     def strip_version_xmlns(self, text):
         """Order of attributes is random, strip version and xmlns"""

--- a/tests/src/python/test_qgsserver_accesscontrol.py
+++ b/tests/src/python/test_qgsserver_accesscontrol.py
@@ -137,17 +137,14 @@ class TestQgsServerAccessControl(QgsServerTestBase):
 
     @classmethod
     def setUpClass(cls):
-        """Run before all tests"""
-        cls._server = QgsServer()
+
+        super().setUpClass()
+
+        cls._server = cls.server  # TODO: drop this
         cls._execute_request("")
-        cls._server_iface = cls._server.serverInterface()
+        cls._server_iface = cls.server.serverInterface()
         cls._accesscontrol = RestrictedAccessControl(cls._server_iface)
         cls._server_iface.registerAccessControl(cls._accesscontrol, 100)
-
-    @classmethod
-    def tearDownClass(cls):
-        """Run after all tests"""
-        del cls._server
 
     def setUp(self):
         super().setUp()

--- a/tests/src/python/test_qgsserver_api.py
+++ b/tests/src/python/test_qgsserver_api.py
@@ -86,7 +86,7 @@ class QgsServerAPIUtilsTest(QgsServerTestBase):
         """Test published WMS CRSs"""
 
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         crss = QgsServerApiUtils.publishedCrsList(project)
         self.assertTrue('http://www.opengis.net/def/crs/OGC/1.3/CRS84' in crss)
         self.assertTrue(
@@ -260,8 +260,8 @@ class QgsServerAPITestBase(QgsServerTestBase):
         self.server.handleRequest(request, response, project)
         result = bytes(response.body()).decode(
             'utf8') if reference_file.endswith('html') else self.dump(response)
-        path = os.path.join(unitTestDataPath(
-            'qgis_server'), subdir, reference_file)
+        path = os.path.join(self.temporary_path,
+                            'qgis_server', subdir, reference_file)
         if self.regeregenerate_api_reference:
             # Try to change timestamp
             try:
@@ -394,7 +394,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
             'http://server.qgis.org/wfs3.json', {'Accept': 'text/html'}, 'application/json')
         # Alias request (we ask for json but we get geojson)
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         self.compareContentType(
             'http://server.qgis.org/wfs3/collections/testlayer%20èé/items?bbox=8.203495,44.901482,8.203497,44.901484',
             {'Accept': 'application/json'}, 'application/geo+json',
@@ -430,7 +430,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
         request = QgsBufferServerRequest(
             'http://server.qgis.org/wfs3/api.openapi3')
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         self.compareApi(request, project, 'test_wfs3_api_project.json')
 
     def test_wfs3_conformance(self):
@@ -457,7 +457,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
         request = QgsBufferServerRequest(
             'http://server.qgis.org/wfs3/collections.json')
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         self.compareApi(request, project, 'test_wfs3_collections_project.json')
 
     def test_wfs3_collections_html(self):
@@ -465,7 +465,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
         request = QgsBufferServerRequest(
             'http://server.qgis.org/wfs3/collections.html')
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         self.compareApi(request, project, 'test_wfs3_collections_project.html')
 
     def test_wfs3_collections_content_type(self):
@@ -475,7 +475,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
             'http://server.qgis.org/wfs3/collections')
         request.setHeader('Accept', 'text/html')
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         response = QgsBufferServerResponse()
         self.server.handleRequest(request, response, project)
         self.assertEqual(response.headers()['Content-Type'], 'text/html')
@@ -489,7 +489,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
     def test_wfs3_collection_json(self):
         """Test WFS3 API collection"""
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         request = QgsBufferServerRequest(
             'http://server.qgis.org/wfs3/collections/testlayer%20èé')
         self.compareApi(request, project,
@@ -518,7 +518,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
     def test_wfs3_collection_html(self):
         """Test WFS3 API collection"""
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         request = QgsBufferServerRequest(
             'http://server.qgis.org/wfs3/collections/testlayer%20èé.html')
         self.compareApi(request, project,
@@ -532,7 +532,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
     def test_wfs3_collection_items(self):
         """Test WFS3 API items"""
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         request = QgsBufferServerRequest(
             'http://server.qgis.org/wfs3/collections/testlayer%20èé/items')
         self.compareApi(request, project,
@@ -541,7 +541,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
     def test_wfs3_collection_items_html(self):
         """Test WFS3 API items"""
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         request = QgsBufferServerRequest(
             'http://server.qgis.org/wfs3/collections/testlayer%20èé/items.html')
         self.compareApi(request, project,
@@ -550,7 +550,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
     def test_wfs3_collection_items_crs(self):
         """Test WFS3 API items with CRS"""
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         encoded_crs = parse.quote(
             'http://www.opengis.net/def/crs/EPSG/9.6.2/3857', safe='')
         request = QgsBufferServerRequest(
@@ -561,8 +561,8 @@ class QgsServerAPITest(QgsServerAPITestBase):
     def test_wfs3_collection_items_as_areas_crs_4326(self):
         """Test WFS3 API items with CRS"""
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') +
-                     '/test_project_wms_grouped_nested_layers.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server',
+                                  'test_project_wms_grouped_nested_layers.qgs'))
         encoded_crs = parse.quote(
             'http://www.opengis.net/def/crs/EPSG/9.6.2/4326', safe='')
         request = QgsBufferServerRequest(
@@ -573,8 +573,8 @@ class QgsServerAPITest(QgsServerAPITestBase):
     def test_wfs3_collection_items_as_areas_crs_3857(self):
         """Test WFS3 API items with CRS"""
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') +
-                     '/test_project_wms_grouped_nested_layers.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server',
+                                  'test_project_wms_grouped_nested_layers.qgs'))
         encoded_crs = parse.quote(
             'http://www.opengis.net/def/crs/EPSG/9.6.2/3857', safe='')
         request = QgsBufferServerRequest(
@@ -585,7 +585,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
     def test_invalid_args(self):
         """Test wrong args"""
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         request = QgsBufferServerRequest(
             'http://server.qgis.org/wfs3/collections/testlayer%20èé/items?limit=-1')
         response = QgsBufferServerResponse()
@@ -605,7 +605,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
     def test_wfs3_collection_items_limit(self):
         """Test WFS3 API item limits"""
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         request = QgsBufferServerRequest(
             'http://server.qgis.org/wfs3/collections/testlayer%20èé/items?limit=1')
         self.compareApi(
@@ -614,7 +614,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
     def test_wfs3_collection_items_limit_offset(self):
         """Test WFS3 API offset"""
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         request = QgsBufferServerRequest(
             'http://server.qgis.org/wfs3/collections/testlayer%20èé/items?limit=1&offset=1')
         self.compareApi(
@@ -637,7 +637,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
     def test_wfs3_collection_items_bbox(self):
         """Test WFS3 API bbox"""
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         request = QgsBufferServerRequest(
             'http://server.qgis.org/wfs3/collections/testlayer%20èé/items?bbox=8.203495,44.901482,8.203497,44.901484')
         self.compareApi(request, project,
@@ -1087,7 +1087,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
     def test_wfs3_field_filters(self):
         """Test field filters"""
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         # Check not published
         response = QgsBufferServerResponse()
         request = QgsBufferServerRequest(
@@ -1104,7 +1104,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
     def test_wfs3_sorting(self):
         """Test sorting"""
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         # Check not published
         response = QgsBufferServerResponse()
         request = QgsBufferServerRequest(
@@ -1133,7 +1133,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
     def test_wfs3_collection_items_properties(self):
         """Test WFS3 API items"""
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
 
         # Invalid request
         request = QgsBufferServerRequest(
@@ -1171,7 +1171,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
     def test_wfs3_field_filters_star(self):
         """Test field filters"""
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         request = QgsBufferServerRequest(
             'http://server.qgis.org/wfs3/collections/layer1_with_short_name/items?name=tw*')
         response = self.compareApi(request, project,
@@ -1181,7 +1181,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
     def test_wfs3_excluded_attributes(self):
         """Test excluded attributes"""
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         request = QgsBufferServerRequest(
             'http://server.qgis.org/wfs3/collections/exclude_attribute/items/0.geojson')
         response = self.compareApi(
@@ -1192,7 +1192,7 @@ class QgsServerAPITest(QgsServerAPITestBase):
         """Test exceptions for invalid fids"""
 
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         request = QgsBufferServerRequest(
             'http://server.qgis.org/wfs3/collections/exclude_attribute/items/123456.geojson')
         response = QgsBufferServerResponse()
@@ -1880,7 +1880,7 @@ class QgsServerOgcAPITest(QgsServerAPITestBase):
         """Test OGC API Handler"""
 
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         request = QgsBufferServerRequest(
             'http://server.qgis.org/wfs3/collections/testlayer%20èé/items?limit=-1')
         response = QgsBufferServerResponse()
@@ -1984,7 +1984,7 @@ class QgsServerOgcAPITest(QgsServerAPITestBase):
         """Test OGC API Handler content types"""
 
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         request = QgsBufferServerRequest(
             'http://server.qgis.org/api3/handlerthree?value1=9.5')
         response = QgsBufferServerResponse()
@@ -2035,7 +2035,7 @@ class QgsServerOgcAPITest(QgsServerAPITestBase):
         """Test OGC API Handler exception"""
 
         project = QgsProject()
-        project.read(unitTestDataPath('qgis_server') + '/test_project_api.qgs')
+        project.read(os.path.join(self.temporary_path, 'qgis_server', 'test_project_api.qgs'))
         request = QgsBufferServerRequest('')
         response = QgsBufferServerResponse()
 

--- a/tests/src/python/test_qgsserver_cachemanager.py
+++ b/tests/src/python/test_qgsserver_cachemanager.py
@@ -153,7 +153,7 @@ class TestQgsServerCacheManager(QgsServerTestBase):
             data = data.encode('utf-8')
         request = QgsBufferServerRequest(qs, requestMethod, {}, data)
         response = QgsBufferServerResponse()
-        cls._server.handleRequest(request, response)
+        cls.server.handleRequest(request, response)
         headers = []
         rh = response.headers()
         rk = sorted(rh.keys())
@@ -163,20 +163,13 @@ class TestQgsServerCacheManager(QgsServerTestBase):
 
     @classmethod
     def setUpClass(cls):
-        """Run before all tests"""
-        cls._app = QgsApplication([], False)
-        cls._server = QgsServer()
+
+        super().setUpClass()
+
         cls._handle_request("")
-        cls._server_iface = cls._server.serverInterface()
+        cls._server_iface = cls.server.serverInterface()
         cls._servercache = PyServerCache(cls._server_iface)
         cls._server_iface.registerServerCache(cls._servercache, 100)
-
-    @classmethod
-    def tearDownClass(cls):
-        """Run after all tests"""
-        # cls._servercache.deleteCachedDocuments(None)
-        del cls._server
-        cls._app.exitQgis
 
     def _result(self, data):
         headers = {}
@@ -191,7 +184,7 @@ class TestQgsServerCacheManager(QgsServerTestBase):
     def _execute_request(self, qs, requestMethod=QgsServerRequest.GetMethod, data=None):
         request = QgsBufferServerRequest(qs, requestMethod, {}, data)
         response = QgsBufferServerResponse()
-        self._server.handleRequest(request, response)
+        self.server.handleRequest(request, response)
         headers = []
         rh = response.headers()
         rk = sorted(rh.keys())
@@ -199,16 +192,8 @@ class TestQgsServerCacheManager(QgsServerTestBase):
             headers.append(("%s: %s" % (k, rh[k])).encode('utf-8'))
         return b"\n".join(headers) + b"\n\n", bytes(response.body())
 
-    def setUp(self):
-        """Create the server instance"""
-        self.fontFamily = QgsFontUtils.standardTestFontFamily()
-        QgsFontUtils.loadStandardTestFonts(['All'])
-
-        d = unitTestDataPath('qgis_server_accesscontrol') + '/'
-        self._project_path = os.path.join(d, "project.qgs")
-
     def test_getcapabilities(self):
-        project = self._project_path
+        project = self.projectPath
         assert os.path.exists(project), "Project file not found: " + project
 
         # without cache
@@ -281,7 +266,7 @@ class TestQgsServerCacheManager(QgsServerTestBase):
         self.assertTrue(cacheManager.deleteCachedDocuments(None), 'deleteCachedDocuments does not return True')
 
     def test_getcontext(self):
-        project = self._project_path
+        project = self.projectPath
         assert os.path.exists(project), "Project file not found: " + project
 
         # without cache
@@ -301,7 +286,7 @@ class TestQgsServerCacheManager(QgsServerTestBase):
         self.assertEqual(len(filelist), 0, 'All files in cache are not deleted ')
 
     def test_describefeaturetype(self):
-        project = self._project_path
+        project = self.projectPath
         assert os.path.exists(project), "Project file not found: " + project
 
         qs = "?" + "&".join(["%s=%s" % i for i in list({
@@ -327,7 +312,7 @@ class TestQgsServerCacheManager(QgsServerTestBase):
         self.assertEqual(len(filelist), 0, 'All files in cache are not deleted ')
 
     def test_getlegendgraphic(self):
-        project = self._project_path
+        project = self.projectPath
         assert os.path.exists(project), "Project file not found: " + project
 
         qs = "?" + "&".join(["%s=%s" % i for i in list({
@@ -368,7 +353,7 @@ class TestQgsServerCacheManager(QgsServerTestBase):
         self.assertEqual(len(filelist), 0, 'All images in cache are not deleted ')
 
     def test_gettile(self):
-        project = self._project_path
+        project = self.projectPath
         assert os.path.exists(project), "Project file not found: " + project
 
         qs = "?" + "&".join(["%s=%s" % i for i in list({
@@ -486,7 +471,7 @@ class TestQgsServerCacheManager(QgsServerTestBase):
         self.assertEqual(len(filelist), 0, 'All images in cache are not deleted ')
 
     def test_gettile_invalid_parameters(self):
-        project = self._project_path
+        project = self.projectPath
         assert os.path.exists(project), "Project file not found: " + project
 
         qs = "?" + "&".join(["%s=%s" % i for i in list({

--- a/tests/src/python/test_qgsserver_locale_override.py
+++ b/tests/src/python/test_qgsserver_locale_override.py
@@ -40,28 +40,16 @@ class TestQgsServerWMSLocaleOverride(TestQgsServerWMSTestBase):
     # Set to True to re-generate reference files for this class
     regenerate_reference = False
 
-    def setUp(self):
-        """Create the server instance"""
-        self.fontFamily = QgsFontUtils.standardTestFontFamily()
-        QgsFontUtils.loadStandardTestFonts(['All'])
-
-        self.testdata_path = unitTestDataPath('qgis_server') + '/'
-
-        d = unitTestDataPath('qgis_server') + '/'
-        self.projectPath = os.path.join(d, "project.qgs")
-
-        # Clean env just to be sure
-        env_vars = ['QUERY_STRING', 'QGIS_PROJECT_FILE']
-        for ev in env_vars:
-            try:
-                del os.environ[ev]
-            except KeyError:
-                pass
+    @classmethod
+    def setUpClass(self):
 
         os.environ['QGIS_SERVER_OVERRIDE_SYSTEM_LOCALE'] = 'EN_us'
         os.environ['QGIS_SERVER_SHOW_GROUP_SEPARATOR'] = '0'
 
-        self.server = QgsServer()
+        super().setUpClass()
+
+        # d = os.path.join(self.temporary_path, 'qgis_server')
+        # self.projectPath = os.path.join(d, "project.qgs")
 
     def testGetFeatureInfoThousandSeparator(self):
 

--- a/tests/src/python/test_qgsserver_wms_dimension.py
+++ b/tests/src/python/test_qgsserver_wms_dimension.py
@@ -37,7 +37,7 @@ class TestQgsServerWMSDimension(TestQgsServerWMSTestBase):
 
     def setUp(self):
         super().setUp()
-        self.testdata_path = unitTestDataPath("qgis_server_accesscontrol")
+        self.testdata_path = os.path.join(self.temporary_path, "qgis_server_accesscontrol")
 
         self.projectPath = os.path.join(self.testdata_path, 'project_with_dimensions.qgs')
         self.assertTrue(os.path.isfile(self.projectPath), 'Could not find project file "{}"'.format(self.projectPath))

--- a/tests/src/python/test_qgsserver_wms_getmap_ignore_bad_layers.py
+++ b/tests/src/python/test_qgsserver_wms_getmap_ignore_bad_layers.py
@@ -45,9 +45,10 @@ class TestQgsServerWMSGetMapIgnoreBadLayers(QgsServerTestBase):
 
     # regenerate_reference = True
 
-    def setUp(self):
+    @classmethod
+    def setUpClass(cls):
         os.environ['QGIS_SERVER_IGNORE_BAD_LAYERS'] = 'true'
-        super().setUp()
+        super().setUpClass()
 
     def test_wms_getmap_datasource_error_ignore(self):
         """Must NOT throw a server exception if datasource if not available and QGIS_SERVER_IGNORE_BAD_LAYERS is set"""
@@ -72,7 +73,7 @@ class TestQgsServerWMSGetMapIgnoreBadLayers(QgsServerTestBase):
 
         r, h = self._result(self._execute_request(qs))
 
-        self.assertFalse('ServerException' in str(r))
+        self.assertFalse('ServerException' in str(r), 'Unexpected ServerException ' + str(r))
 
 
 if __name__ == '__main__':

--- a/tests/src/python/test_qgsserver_wms_getmap_size_server.py
+++ b/tests/src/python/test_qgsserver_wms_getmap_size_server.py
@@ -36,10 +36,13 @@ class TestQgsServerWMSGetMapSizeServer(QgsServerTestBase):
     # Set to True to re-generate reference files for this class
     regenerate_reference = False
 
-    def setUp(self):
+    @classmethod
+    def setUpClass(self):
         os.environ['QGIS_SERVER_WMS_MAX_WIDTH'] = '3000'
         os.environ['QGIS_SERVER_WMS_MAX_HEIGHT'] = '3000'
-        super(TestQgsServerWMSGetMapSizeServer, self).setUp()
+        super().setUpClass()
+
+    def setUp(self):
         self.project = os.path.join(self.testdata_path, "test_project_with_size.qgs")
         self.expected_too_big = self.strip_version_xmlns(b'<ServiceExceptionReport version="1.3.0" xmlns="http://www.opengis.net/ogc">\n <ServiceException code="InvalidParameterValue">The requested map size is too large</ServiceException>\n</ServiceExceptionReport>\n')
 

--- a/tests/src/python/test_qgsserver_wms_getprint_maptheme.py
+++ b/tests/src/python/test_qgsserver_wms_getprint_maptheme.py
@@ -37,12 +37,10 @@ class PyQgsServerWMSGetPrintMapTheme(QgsServerTestBase):
     @classmethod
     def setUpClass(cls):
 
-        cls.tmp_dir = QTemporaryDir()
-        shutil.copyfile(os.path.join(unitTestDataPath('qgis_server'), 'test_project_mapthemes.qgs'), os.path.join(cls.tmp_dir.path(), 'test_project_mapthemes.qgs'))
-        shutil.copyfile(os.path.join(unitTestDataPath('qgis_server'), 'test_project_mapthemes.gpkg'), os.path.join(cls.tmp_dir.path(), 'test_project_mapthemes.gpkg'))
+        super().setUpClass()
 
         project = QgsProject()
-        assert (project.read(os.path.join(cls.tmp_dir.path(), 'test_project_mapthemes.qgs')))
+        assert (project.read(os.path.join(cls.temporary_path, 'qgis_server', 'test_project_mapthemes.qgs')))
 
         cls.project = project
         cls.polygon = 'POLYGON((7.09769689415099325 44.92867722467413216, 7.37818833364500737 44.92867722467413216, 7.37818833364500737 45.0714498943264914, 7.09769689415099325 45.0714498943264914, 7.09769689415099325 44.92867722467413216))'

--- a/tests/src/python/test_qgsserver_wmts.py
+++ b/tests/src/python/test_qgsserver_wmts.py
@@ -68,7 +68,7 @@ class TestQgsServerWMTS(QgsServerTestBase):
             reference_name = 'wmts_' + request.lower()
 
         reference_name += '.txt'
-        reference_path = self.testdata_path + reference_name
+        reference_path = os.path.join(self.testdata_path, reference_name)
 
         self.store_reference(reference_path, response)
         f = open(reference_path, 'rb')
@@ -98,7 +98,7 @@ class TestQgsServerWMTS(QgsServerTestBase):
             reference_name = 'wmts_' + request.lower()
 
         reference_name += '.txt'
-        reference_path = self.testdata_path + reference_name
+        reference_path = os.path.join(self.testdata_path, reference_name)
 
         self.store_reference(reference_path, response)
         f = open(reference_path, 'rb')

